url/urlsearchparams.go

```diff
@@ -0,0 +1,424 @@
+		}
+
+		u := toURL(r, call.This)
+		name := call.Arguments[0].String()
+		if len(call.Arguments) > 1 {
+			value := call.Arguments[1].String()
+			arr := searchParams{}
+			for _, v := range u.searchParams {
+				if v.name != name {
+					arr = append(arr, v)
+				} else {
+					subArr := []string{}
+					for _, val := range v.value {
+						if val != value {
+							subArr = append(subArr, val)
+						}
+					}
+					if len(subArr) > 0 {
+						arr = append(arr, searchParam{name: name, value: subArr})
+					}
+				}
+			}
+			u.searchParams = arr
+		} else {
+			arr := searchParams{}
+			for _, v := range u.searchParams {
+				if v.name != name {
+					arr = append(arr, v)
+				}
+			}
+			u.searchParams = arr
+		}
+		u.syncSearchParams()
+
+		return goja.Undefined()
+	}))
+
+	p.Set("entries", r.ToValue(func(call goja.FunctionCall) goja.Value {
+		u := toURL(r, call.This)
+		entries := [][]string{}
+		for _, sp := range u.searchParams {
+			entries = append(entries, []string{sp.name, strings.Join(sp.value, ",")})
+		}
+
+		return r.ToValue(entries)
+	}))
+
+	p.Set("forEach", r.ToValue(func(call goja.FunctionCall) goja.Value {
+		if len(call.Arguments) != 1 {
+			panic(newInvalidArgsError(r))
+		}
+
+		u := toURL(r, call.This)
+		if fn, ok := goja.AssertFunction(call.Arguments[0]); ok {
+			for _, pair := range u.searchParams {
+				// name, value, searchParams
+				for _, v := range pair.value {
+					query := u.url.RawQuery
+					_, err := fn(
+						nil,
+						r.ToValue(pair.name),
+						r.ToValue(v),
+						r.ToValue(query),
+					)
+
+					if err != nil {
+						panic(err)
+					}
+				}
+			}
+		}
+
+		return goja.Undefined()
+	}))
+
+	p.Set("get", r.ToValue(func(call goja.FunctionCall) goja.Value {
+		if len(call.Arguments) == 0 {
+			panic(newMissingArgsError(r, `The "name" argument must be specified`))
+		}
+
+		p := call.Arguments[0]
+		e := p.Export()
+		if n, ok := e.(string); ok {
+			u := toURL(r, call.This)
+			vals, _ := u.getValues(n)
+			if len(vals) > 0 {
+				return r.ToValue(vals[0])
+			}
+		}
+
+		return goja.Null()
+	}))
+
+	p.Set("getAll", r.ToValue(func(call goja.FunctionCall) goja.Value {
+		if len(call.Arguments) == 0 {
+			panic(newMissingArgsError(r, `The "name" argument must be specified`))
+		}
+
+		p := call.Arguments[0]
+		e := p.Export()
+		if n, ok := e.(string); ok {
+			u := toURL(r, call.This)
+			vals, _ := u.getValues(n)
+			if len(vals) > 0 {
+				return r.ToValue(vals)
+			}
+		}
+
+		return goja.Null()
+	}))
+
+	p.Set("has", r.ToValue(func(call goja.FunctionCall) goja.Value {
+		if len(call.Arguments) == 0 {
+			panic(newMissingArgsError(r, `The "name" argument must be specified`))
+		}
+
+		p := call.Arguments[0]
+		e := p.Export()
+		if n, ok := e.(string); ok {
+			u := toURL(r, call.This)
+			vals, contained := u.getValues(n)
+			if len(call.Arguments) > 1 {
+				for _, v := range vals {
+					cmp := call.Arguments[1].String()
```

Comment:
This should be hoisted outside the loop. Even though String() may have side effects, checking with nodejs shows it's only called once.